src/math.jl

```diff
@@ -35,6 +35,21 @@ Base.:^(l::AbstractQuantity{T,D}, r::Number) where {T,R,D<:AbstractDimensions{R}
         return new_quantity(typeof(l), ustrip(l)^val_pow, dimension(l)^dim_pow)
     end
 
+@inline Base.literal_pow(::typeof(^), l::AbstractQuantity, ::Val{p}) where {p} = l^p
+
+# Special forms for small integer powers:
+# https://github.com/JuliaLang/julia/blob/b99f251e86c7c09b957a1b362b6408dbba106ff0/base/intfuncs.jl#L332
+for (p, ex) in [
+    (0, :(one(l))),
+    (1, :(l)),
+    (2, :(l * l)),
+    (3, :(l * l * l)),
+    (-1, :(inv(l))),
+    (-2, :((i=inv(l); i*i)))
+]
+    @eval @inline Base.literal_pow(::typeof(^), l::AbstractQuantity, ::Val{$p}) = $ex
+end
```

Comment:
```suggestion
@inline function Base.literal_pow(::typeof(^), l::AbstractQuantity, v::Val)
    new_quantity(typeof(l), Base.literal_pow(^, ustrip(l), v), Base.literal_pow(^, dimension(l), v))
end
```
This seems to do just as well in benchmarking 🙂

Reply:
How are you benchmarking it? For some reason I get worse performance on the broadcasting benchmark:

1. With explicit expansion like in `intfuncs.jl`:

```julia
julia> f(x) = x^2 * 0.9 + 0.9 * x / 0.1 * x;

julia> @btime f.(x) setup=(x = QuantityArray(randn(100000), u"km/s"));
  97.250 μs (2 allocations: 781.30 KiB)
```

2. With the generic `@inline Base.literal_pow(::typeof(^), l::AbstractQuantity, ::Val{p}) where {p} = new_quantity(typeof(l), Base.literal_pow(^, ustrip(l), Val(p)), Base.literal_pow(^, dimension(l), Val(p)))`:

```julia
julia> @btime f.(x) setup=(x = QuantityArray(randn(100000), u"km/s"));
  1.130 ms (2 allocations: 781.30 KiB)
```

But I'll also note that if I change `f(x)` to use `x^4` (and similar for the right term), avoiding the `literal_pow`, then I also get similarly poor performance for 1.

I wonder why this is. Maybe the general power law method for quantities is too complex, and messes with the compiler inlining, thus forcing the broadcasted expression to do dimension checking for every number? (Whereas for `x * x * x` the compiler can easily inline the dimension calculations).

Maybe I need to add a power law method for `Core.Const()`??

Reply:
Code in question:

```julia
_pow(l::AbstractDimensions{R}, r::R) where {R} = map_dimensions(Base.Fix1(*, r), l)
Base.:^(l::AbstractDimensions{R}, r::Number) where {R} = _pow(l, tryrationalize(R, r))
Base.:^(l::AbstractQuantity{T,D}, r::Integer) where {T,R,D<:AbstractDimensions{R}} = new_quantity(typeof(l), ustrip(l)^r, dimension(l)^r)
Base.:^(l::AbstractQuantity{T,D}, r::Number) where {T,R,D<:AbstractDimensions{R}} =
    let dim_pow = tryrationalize(R, r), val_pow = convert(T, dim_pow)
        # Need to ensure we take the numerical power by the rationalized quantity:
        return new_quantity(typeof(l), ustrip(l)^val_pow, dimension(l)^dim_pow)
    end
```

Reply:
You're right, I don't know how that didn't show locally.

I also thought it was inlining, but I realized there's actually a semantics difference with how the dimensions were getting transformed. In the old code, the multiplication would end up being unrolled into repeated addition, but with the generic version it stays as multiplication.

Although it unfortunately doesn't result in any decrease in code complexity, the right way to handle this seems to be to keep the generic `literal_pow` method for quantities, but actually add a specialization for dimensions. I've implemented this in #36 (on my machine, this actually gets a speedup on your benchmark over even the original code!)

Reply:
(That being said, afaiu we wouldn't usually expect repeated addition to be an improvement over multiplication... but maybe it plays better with inlining somehow)

Reply:
In looking at the code trace for the higher order literal pow, it converts from `Core.Const` to `Int64` when it reaches the `^`. So I wonder if we should use the `map_domensions` call directly in the literal pow?

Reply:
I just experimented with adding the following line to #36:
```julia
Base.literal_pow(::typeof(^), l::AbstractDimensions, ::Val{4}) = map_dimensions(Base.Fix1(*, 4), l)
```
but it doesn't seem to help much if at all. However,
```julia
Base.literal_pow(::typeof(^), l::AbstractDimensions, ::Val{4}) = map_dimensions(+
, l, l, l, l)
```
does help a lot. But I wasn't able to generically get the compiler to const-prop for higher powers, something to get to the bottom of...

For now I added a simple benchmark to #36